src/backend/distributed/planner/local_distributed_join_planner.c

```diff
@@ -0,0 +1,529 @@
+	List *restrictClauseList = get_all_actual_clauses(baseRestrictionList);
+	if (ContainsFalseClause(restrictClauseList))
+	{
+		/* If there is a WHERE FALSE, we consider it as a constant filter. */
+		return true;
+	}
+	List *rteEqualityColumnsNos =
+		FetchEqualityAttrNumsForRTE((Node *) restrictClauseList);
+
+	List *uniqueIndexColumnsList = ExecuteFunctionOnEachTableIndex(rangeTableEntry->relid,
+																   GetAllUniqueIndexes);
+	IndexColumns *indexColumns = NULL;
+	foreach_ptr(indexColumns, uniqueIndexColumnsList)
+	{
+		List *uniqueIndexColumnNos = indexColumns->indexColumnNos;
+		if (FirstIntListContainsSecondIntList(rteEqualityColumnsNos,
+											  uniqueIndexColumnNos))
+		{
+			return true;
+		}
+	}
+	return false;
+}
+
+
+/*
+ * FirstIntListContainsSecondIntList returns true if the first int List
+ * contains every element of the second int List.
+ */
+static bool
+FirstIntListContainsSecondIntList(List *firstIntList, List *secondIntList)
+{
+	int curInt = 0;
+	foreach_int(curInt, secondIntList)
+	{
+		if (!list_member_int(firstIntList, curInt))
+		{
+			return false;
+		}
+	}
+	return true;
+}
+
+
+/*
+ * GetAllUniqueIndexes adds the given index's column numbers if it is a
+ * unique index.
+ * TODO:: if there is a unique index on a multiple column, then we should
+ * probably return true only if all the columns in the index exist in the filter.
+ */
+static void
+GetAllUniqueIndexes(Form_pg_index indexForm, List **uniqueIndexGroups)
+{
+	if (indexForm->indisunique || indexForm->indisprimary)
+	{
+		IndexColumns *indexColumns = palloc0(sizeof(IndexColumns));
+		List *uniqueIndexes = NIL;
+		for (int i = 0; i < indexForm->indkey.dim1; i++)
+		{
+			uniqueIndexes = list_append_unique_int(uniqueIndexes,
+												   indexForm->indkey.values[i]);
+		}
+		if (list_length(uniqueIndexes) == 0)
+		{
+			return;
+		}
+		indexColumns->indexColumnNos = uniqueIndexes;
+		*uniqueIndexGroups = lappend(*uniqueIndexGroups, indexColumns);
+	}
+}
+
+
+/*
+ * RequiredAttrNumbersForRelation returns the required attribute numbers for
+ * the input RTE relation in order for the planning to succeed.
+ *
+ * The function could be optimized by not adding the columns that only appear
+ * WHERE clause as a filter (e.g., not a join clause).
+ */
+static List *
+RequiredAttrNumbersForRelation(RangeTblEntry *rangeTableEntry,
+							   PlannerRestrictionContext *plannerRestrictionContext)
+{
+	RelationRestriction *relationRestriction =
+		RelationRestrictionForRelation(rangeTableEntry, plannerRestrictionContext);
+
+	if (relationRestriction == NULL)
+	{
+		return NIL;
+	}
+
+	PlannerInfo *plannerInfo = relationRestriction->plannerInfo;
+	Query *queryToProcess = plannerInfo->parse;
+	int rteIndex = relationRestriction->index;
+
+	List *allVarsInQuery = pull_vars_of_level((Node *) queryToProcess, 0);
+
+	List *requiredAttrNumbers = NIL;
+
+	Var *var = NULL;
+	foreach_ptr(var, allVarsInQuery)
+	{
+		if (var->varno == rteIndex)
```

Comment:
hmz, how does this work for subqueries? the same varno will be used for many different relations and subqueries

Reply:
So here we will collect the vars that references level 0. For the variables that are inside subqueries the varno will point to the index in the level they are referencing to. So in that case, it will be equivalent to rteIndex only if it actually referring to the current RTE.

As an example:

```sql
SELECT * FROM local JOIN dist d1 USING(a) JOIN LATERAL ( SELECT * FROM dist where d1.a = 10)l USING(a);
```
Here the inner d1.a's varno will be 2. 

Though here do we need column b while wrapping `dist` table?
```sql
[local] talha@talha:9700-24566=# SELECT COUNT(*) FROM local JOIN dist d1 ON(local.a = d1.b) JOIN LATERAL │
( SELECT * FROM dist where d1.a = 10)l ON(local.a = d1.b);                                               │
DEBUG:  Wrapping relation "dist" "d1" to a subquery                                                      │
DEBUG:  generating subplan 10_1 for subquery SELECT a, b FROM public.dist d1 WHERE (a OPERATOR(pg_catalog│
.=) 10)                                                                                                  │
```

Reply:
Got confused by the name `allVarsInQuery`, should probably be renamed.